qubes-gui-client/src/lib.rs

```diff
@@ -1,7 +1,146 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+//! A client for the Qubes OS GUI protocol.  This client is low-level.
+
+#![forbid(missing_docs)]
+#![forbid(unconditional_recursion)]
+
+use qubes_castable::Castable as _;
+pub use qubes_gui;
+use std::collections::BTreeSet;
+use std::io;
+use std::num::NonZeroU32;
+use std::task::Poll;
+
+mod buffer;
+
+/// The entry-point to the library.
+#[derive(Debug)]
+pub struct Client {
+    vchan: buffer::Vchan<vchan::Vchan>,
+    set: BTreeSet<NonZeroU32>,
+    agent: bool,
+}
+
+impl Client {
+    /// Send a GUI message.  This never blocks; outgoing messages are queued
+    /// until there is space in the vchan.
+    pub fn send<T: qubes_gui::Message>(
+        &mut self,
+        message: &T,
+        window: NonZeroU32,
+    ) -> io::Result<()> {
+        self.send_raw(message.as_bytes(), window, T::kind())
+    }
+
+    /// Raw version of [`Client::send`].  Using [`Client::send`] is preferred
+    /// where possible, as it automatically selects the correct message type.
+    pub fn send_raw(&mut self, message: &[u8], window: NonZeroU32, ty: u32) -> io::Result<()> {
+        let untrusted_len = message.len() as u32;
+        assert_eq!(
+            untrusted_len as usize,
+            message.len(),
+            "Message length must fit in a u32"
+        );
+        let header = qubes_gui::Header {
+            ty,
+            window: window.into(),
+            untrusted_len,
+        };
+        if self.agent {
+            if header.ty == qubes_gui::MSG_CREATE {
+                assert!(
+                    self.set.insert(window),
+                    "Creating window {} already in map!",
+                    window
+                );
+            } else if header.ty == qubes_gui::MSG_DESTROY {
+                assert!(
+                    self.set.remove(&window),
+                    "Trying to delete window {} not in map!",
+                    window
+                );
+            } else {
+                assert!(
+                    self.set.contains(&window),
+                    "Sending message on nonexistant window {}!",
+                    window
+                )
+            }
+        }
+        // FIXME this is slow
+        self.vchan.write(header.as_bytes())?;
+        self.vchan.write(message)?;
+        Ok(())
+    }
+
+    /// Even rawer version of [`Client::send`].  Using [`Client::send`] is
+    /// preferred where possible, as it automatically selects the correct
+    /// message type.  Otherwise, prefer [`Client::send_raw`], which at least
+    /// ensures correct framing.
+    pub fn send_raw_bytes(&mut self, msg: &[u8]) -> io::Result<()> {
+        self.vchan.write(msg)
+    }
+
+    /// Acknowledge an event (as reported by poll(2), epoll(2), or similar).
+    /// Must be called before performing any I/O.
+    pub fn wait(&mut self) {
+        self.vchan.wait()
+    }
```

Comment:
`wait` is a weird name for an ACK function...

Also, I'd consider whether this invariant is something we could take care of at this level.